provider/core/src/serde.rs

```diff
@@ -282,18 +282,34 @@ impl<'a> Deref for SerdeSeDataStructDynRef<'a> {
 
 impl<'data, M> crate::dynutil::UpcastDataPayload<'data, M> for SerdeSeDataStructMarker
 where
-    M: DataMarker<'data>,
+    M: DataMarker,
     for<'a> &'a <M::Yokeable as Yokeable<'a>>::Output: serde::Serialize,
-    M::Cart: IsCovariant<'data>,
 {
     fn upcast(other: DataPayload<'data, M>) -> DataPayload<'data, SerdeSeDataStructMarker> {
-        use crate::data_provider::DataPayloadInner::*;
-        let cart: Rc<dyn SerdeSeDataStruct<'data> + 'data> = match other.inner {
-            RcStruct(yoke) => Rc::from(yoke),
-            Owned(yoke) => Rc::from(yoke),
-            RcBuf(yoke) => Rc::from(yoke),
+        use crate::data_provider::{DataPayloadInner, ErasedCart};
+        let yoke: Yoke<_, ErasedCart> = match other.inner {
+            DataPayloadInner::RcStruct(yoke) => {
+                let rc: Rc<Yoke<_, _>> = Rc::from(yoke);
+                let yoke =
+                    Yoke::attach_to_rc_cart(rc.clone() as Rc<dyn SerdeSeDataStruct<'data> + 'data>);
+                unsafe { yoke.replace_cart(move |_| rc as ErasedCart<'data>) }
```

Comment:
Suggestion, here and below: you don't need to `clone` the Rc:

```suggestion
                let rc: Rc<Yoke<_, _>> = Rc::from(yoke);
                let yoke =
                    Yoke::attach_to_rc_cart(rc as Rc<dyn SerdeSeDataStruct<'data> + 'data>);
                unsafe { yoke.replace_cart(move |c| c as ErasedCart<'data>) }
```

After this, you can change the Rc to Box.

Reply:
You can't cast trait objects like that

Reply:
Because `dyn SerdeSeDataStruct<'data> + 'data` doesn't implement `dyn ErasedDestructor<'data> + 'data` ?

Reply:
No, you cannot cast between one trait object and another trait object. We could make this work by adding another method called `as_rc_erased_destructor()` on `SerdeSeDataStruct` but that's a codesize hazard for something that's a very cheap temporary refcount bump anyway.

Reply:
ok, I'm convinced